nx/lib/nx.ex

```diff
@@ -3083,8 +3105,30 @@ defmodule Nx do
 
   ## Element-wise binary ops
 
+  @allow_complex_type_bin_ops [
+    :add,
+    :subtract,
+    :multiply,
+    :divide,
+    :atan2,
+    :power,
+    :equal,
+    :not_equal,
+    :greater,
+    :less,
+    :greater_equal,
+    :logical_and,
+    :logical_or,
+    :logical_xor
+  ]
+
   defp element_wise_bin_op(left, right, op, fun) do
     type = binary_type(left, right) |> fun.()
+
+    if op not in @allow_complex_type_bin_ops do
```

Comment:
Instead of hard coding it here, I would have a `non_complex_element_wise_bin_op` that raises for complex and then calls this one.